community/neo4j/src/test/java/ConcurrentChangesOnEntitiesTest.java

```diff
@@ -0,0 +1,254 @@
+            {
+                node.addLabel( DynamicLabel.label( "A" ) );
+            }
+        } );
+
+        startAndWait( t1, t2 );
+
+        db.shutdown();
+
+        assertDatabaseConsistent();
+    }
+
+    @Test
+    public void setConcurrentlySamePropertyWithDifferentValuesOnANode() throws Throwable
+    {
+        final long nodeId = initWithNode( db );
+
+        Thread t1 = newThreadForNodeAction( nodeId, new Consumer<Node>()
+        {
+            @Override
+            public void accept( Node node )
+            {
+                node.setProperty( "a", 0.788 );
+            }
+        } );
+
+        Thread t2 = newThreadForNodeAction( nodeId, new Consumer<Node>()
+        {
+            @Override
+            public void accept( Node node )
+            {
+                node.setProperty( "a", new double[]{0.999, 0.77} );
+            }
+        } );
+
+        startAndWait( t1, t2 );
+
+        db.shutdown();
+
+        assertDatabaseConsistent();
+    }
+
+    @Test
+    public void setConcurrentlySamePropertyWithDifferentValuesOnARelationship() throws Throwable
+    {
+        final long relId = initWithRel( db );
+
+        Thread t1 = newThreadForRelationshipAction( relId, new Consumer<Relationship>()
+        {
+            @Override
+            public void accept( Relationship relationship )
+            {
+                relationship.setProperty( "a", 0.788 );
+            }
+        } );
+
+        Thread t2 = newThreadForRelationshipAction( relId, new Consumer<Relationship>()
+        {
+            @Override
+            public void accept( Relationship relationship )
+            {
+                relationship.setProperty( "a", new double[]{0.999, 0.77} );
+            }
+        } );
+
+        startAndWait( t1, t2 );
+
+        db.shutdown();
+
+        assertDatabaseConsistent();
+    }
+
+    private long initWithNode( GraphDatabaseService db )
+    {
+        try ( Transaction tx = db.beginTx() )
+        {
+            Node theNode = db.createNode();
+            long id = theNode.getId();
+            tx.success();
+            return id;
+        }
+
+    }
+
+    private long initWithRel( GraphDatabaseService db )
+    {
+        try ( Transaction tx = db.beginTx() )
+        {
+            Node node = db.createNode();
+            node.setProperty( "a", "prop" );
+            Relationship rel = node.createRelationshipTo( db.createNode(), DynamicRelationshipType.withName( "T" ) );
+            long id = rel.getId();
+            tx.success();
+            return id;
+        }
+    }
+
+    private Thread newThreadForNodeAction( final long nodeId, final Consumer<Node> nodeConsumer )
+    {
+        return new Thread( new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                try ( Transaction tx = db.beginTx() )
+                {
+                    Node node = db.getNodeById( nodeId );
+                    ready.countDown();
```

Comment:
One `CyclicBarrier` can be used in all cases here instead of two `CountDownLatch`